core/src/sumeragi/main_loop.rs

```diff
@@ -430,325 +449,381 @@ impl Sumeragi {
+                                .unpack(|e| self.send_event(e))
+                            {
+                                Ok(committed_block) => {
+                                    self.commit_block(committed_block, voted_block.block_state)
+                                }
+                                Err((mut block, error)) => {
+                                    error!(
+                                        peer_id=%self.peer_id,
+                                        role=%self.role(),
+                                        ?error,
+                                        "Block failed to be committed"
+                                    );
+
+                                    block
+                                        .replace_signatures(prev_signatures, &self.topology)
+                                        .unpack(|e| self.send_event(e))
+                                        .expect("INTERNAL BUG: Failed to replace signatures");
+                                    voted_block.block = block;
+                                    *voting_block = Some(voted_block);
+                                }
+                            }
+                        }
+                        Err(error) => {
+                            error!(
+                                peer_id=%self.peer_id,
+                                role=%self.role(),
+                                ?error,
+                                "Received incorrect signatures"
+                            );
+
+                            *voting_block = Some(voted_block);
+                        }
+                    }
+                } else {
+                    error!(
+                        peer_id=%self.peer_id,
+                        role=%self.role(),
+                        "Peer missing voting block"
+                    );
+                }
+            }
+            (msg, _) => {
+                trace!(
+                    role=%self.role(),
+                    peer_id=%self.peer_id,
+                    ?msg,
+                    "message not handled"
+                );
             }
         }
     }
 
+    /// Commits block if there are enough votes
+    fn try_commit_block<'state>(
+        &mut self,
+        mut voting_block: VotingBlock<'state>,
+        #[cfg_attr(not(debug_assertions), allow(unused_variables))] is_genesis_peer: bool,
+    ) -> Option<VotingBlock<'state>> {
+        assert_eq!(self.role(), Role::ProxyTail);
+
+        let votes_count = voting_block.block.as_ref().signatures().len();
+        if votes_count + 1 >= self.topology.min_votes_for_commit() {
+            voting_block.block.sign(&self.key_pair, &self.topology);
+
+            let committed_block = voting_block
+                .block
+                .commit(&self.topology)
+                .unpack(|e| self.send_event(e))
+                .expect("INTERNAL BUG: Proxy tail failed to commit block");
+
```

Comment:
Maybe instead of unwrapping we can try to commit the block in the `if` statement?
I think it would be more reliable in long term if we decide that to commit block it's not enough to just receive enough signatures.
And proxy tail can add it's signature on `BlockCreated` message.

Reply:
this function is only callable by the `ProxyTail` and `ValidBlock` is guaranteed to have all signatures valid. Consequently, this cannot and must not fail unless the `ProxyTail` is itself faulty